src/main/java/uk/gov/justice/hmpps/prison/service/AdjudicationsService.java

```diff
@@ -264,4 +323,167 @@ private AdjudicationDetail transformToDto(final Adjudication adjudication) {
         final var adjudicationCharges = offenderPartyDetails.get().getCharges();
         return adjudicationCharges.stream().map(c -> c.getOffenceType().getOffenceCode()).collect(Collectors.toList());
     }
+
+    private Adjudication updateAncillaryAdjudicationParties(
+        @NotNull Adjudication adjudication,
+        @NotNull List<Long> victimStaffIds,
+        @NotNull List<String> victimOffenderIds,
+        @NotNull List<String> connectedOffenderIds)
+    {
+        final var victimStaff = victimStaffIds != null ?
+            victimStaffIds.stream()
+                .map(id -> staffUserAccountRepository.findByStaff_StaffId(id)
+                    .orElseThrow(() -> new RuntimeException(format("User not found %s", id))).getStaff()
+                ).collect(Collectors.toList()) : List.<Staff>of();
+        final var victimOffenderBookings = victimOffenderIds != null ?
+            victimOffenderIds.stream()
+                .map(id -> bookingRepository.findByOffenderNomsIdAndBookingSequence(id, 1)
+                    .orElseThrow(() -> new RuntimeException(format("Could not find a current booking for Offender No %s", id)))
+                ).collect(Collectors.toList()) : List.<OffenderBooking>of();
+        final var connectedOffenderBookings = connectedOffenderIds != null ?
+            connectedOffenderIds.stream().map(id -> bookingRepository.findByOffenderNomsIdAndBookingSequence(id, 1)
+                .orElseThrow(() -> new RuntimeException(format("Could not find a current booking for Offender No %s", id)))
+            ).collect(Collectors.toList()) : List.<OffenderBooking>of();
+
+        var remainingAdjudicationParties = ancillaryAdjudicationPartiesToRemain(victimStaff, victimOffenderBookings, connectedOffenderBookings, adjudication);
+        var offenderParty = adjudication.getOffenderParty();
+        adjudication.getParties().clear();
+        adjudication.getParties().add(offenderParty.get());
+        adjudication.getParties().addAll(remainingAdjudicationParties);
+        // There is a database trigger AGENCY_INCIDENT_PARTIES_T1 intended to maintain the uniqueness of offender bookings
+        // If we do not flush at this point then hibernate batch SQL generated may add a new row before deleting an old
+        // one which violates the trigger - so we need to flush here to ensure ordering.
+        entityManager.flush();
+
+        var newAdjudicationParties = newAncillaryAdjudicationParties(victimStaff, victimOffenderBookings, connectedOffenderBookings, adjudication);
```

Comment:
So originally I had something exactly like this - but the trigger got in the way. The only way I've managed to get it working so far is to:
1) Remove the adjudication parties for which there is no match with the new ones passed in by the dto
2) flush
3) Add the new adjudication parties one at a time and flush after each is added
And so because of this I need to work out which remain first and add those. Then I need to work out which to add and then add those. If that makes sense?